services/google_scraping_service/parser.go

```diff
@@ -0,0 +1,75 @@
+package google_scraping_service
+
+import (
+	"net/http"
+
+	"github.com/PuerkitoBio/goquery"
+)
+
+type GoogleResponseParser struct {
+	GoogleResponse *http.Response
+}
+
+type ScrapingResult struct {
+	LinksCount             int
+	NonAdwordsCount        int
+	NonAdwordLinks         []string
+	TopPostionAdwordsCount int
+	TopPositionAdwordLinks []string
+	TotalAdwordsCount      int
+}
+
+func (g *GoogleResponseParser) ParseGoogleResponse() (ScrapingResult, error) {
+	doc, err := goquery.NewDocumentFromReader(g.GoogleResponse.Body)
+	if err != nil {
+		return ScrapingResult{}, err
+	}
+
+	scrapingResult := ScrapingResult{
+		LinksCount:             g.countLinks(doc),
+		NonAdwordsCount:        g.countNonAdwords(doc),
+		NonAdwordLinks:         g.fetchNonAdwordLinks(doc),
+		TopPostionAdwordsCount: g.countTopPositionAdwords(doc),
+		TopPositionAdwordLinks: g.fetchTopPositionAdwordLinks(doc),
+		TotalAdwordsCount:      g.countTotalAdwords(doc),
```

Comment:
It might be good as a nice-to-have improvement to look into goroutines to parse this asynchronously, now it's very linear

Reply:
Improved in 89f24867d743bdfa366afba6ae2f85715033ad50

Reply:
Found an issue when working with background job. Some result not parsing properly, so I will go with the linear style for now and improve it at the end or maybe in the separated branch.

Reply:
> Found an issue when working with background job. Some result not parsing properly, so I will go with the linear style for now and improve it at the end or maybe in the separated branch.

Fixed this issue. It is not about goroutine stuff. The issue caused from my inconsistent css selector, so now I went back to goroutine way.